### PR TITLE
multi-introducer (2788)

### PR DESCRIPTION
**Pull request by meejah**
I believe this contains the same changes as #284 except with proper commit attribution.

This particular branch is `rebase-multiintro-leif-david` rebased onto master; if you want to compare my attribution branch to David's original pull request, use `rebase-multiintro-leif-david` as they have the same branch-point.

### CODE REVIEW

**Comment by meejah** [comment]
Okay, this should be good-to-go, after re-doing some incorrect rebase conflict resolutions (ask me about forgetting `git rerere` information ;). I think that's a spurious failure in travis?

File: src/allmydata/client.py [hunk 1]
```diff
@@ -170,17 +175,63 @@ def _sequencer(self):
...
-        introducer_cache_filepath = FilePath(os.path.join(self.basedir, "private", "introducer_cache.yaml"))
-        ic = IntroducerClient(self.tub, self.introducer_furl,
-                              self.nickname,
-                              str(allmydata.__full_version__),
-                              str(self.OLDEST_SUPPORTED_VERSION),
-                              self.get_app_versions(),
-                              self._sequencer, introducer_cache_filepath)
-        self.introducer_client = ic
-        ic.setServiceParent(self)
+    def old_introducer_config_compatiblity(self):
```
Comment (daira) [inline src/allmydata/client.py]: Typo in "compatibility", also, this name is not very clear to me (method names should usually be verbs).

File: src/allmydata/client.py [hunk 2]
```diff
@@ -170,17 +175,63 @@ def _sequencer(self):
...
+            exists = False
+            self.connections_config = { 'servers' : {},
+                                        'introducers' : {},
+            }
+            connections_filepath.setContent(yaml.safe_dump(self.connections_config))
+
+        self.old_introducer_config_compatiblity()
+        introducers = self.connections_config['introducers']
+        for nickname in introducers:
+            introducer_cache_filepath = FilePath(os.path.join(self.basedir, "private", nickname))
```
Comment (daira) [inline src/allmydata/client.py]: Misleading name? This is not just a cache.

File: src/allmydata/test/test_multi_introducers.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        # prepare tahoe.cfg
+        c = open(os.path.join(self.basedir,"tahoe.cfg"), "w")
+        config = {}
+        write_node_config(c, config)
+        fake_furl = "furl1"
+        c.write("[client]\n")
+        c.write("introducer.furl = %s\n" % fake_furl)
+        c.close()
+
+        # prepare "basedir/private/connections.yml
```
Comment (daira) [inline src/allmydata/test/test_multi_introducers.py]: misplaced "

**Pull request closed without merging**